platform/security/rest/security-rest-cxfwrapper/src/main/java/org/codice/ddf/cxf/oauth/OAuthSecurityImpl.java

```diff
@@ -0,0 +1,256 @@
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static javax.ws.rs.core.HttpHeaders.AUTHORIZATION;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static org.apache.http.HttpStatus.SC_OK;
+import static org.codice.gsonsupport.GsonTypeAdapters.MAP_STRING_TO_OBJECT_TYPE;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.nimbusds.jose.util.DefaultResourceRetriever;
+import com.nimbusds.jose.util.ResourceRetriever;
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.SignedJWT;
+import com.nimbusds.oauth2.sdk.ParseException;
+import com.nimbusds.oauth2.sdk.token.BearerAccessToken;
+import com.nimbusds.openid.connect.sdk.op.OIDCProviderMetadata;
+import ddf.security.Subject;
+import ddf.security.SubjectUtils;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URL;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Map;
+import javax.ws.rs.core.Form;
+import org.apache.commons.io.IOUtils;
+import org.apache.cxf.jaxrs.client.Client;
+import org.apache.cxf.jaxrs.client.WebClient;
+import org.apache.http.HttpStatus;
+import org.apache.logging.log4j.util.Strings;
+import org.codice.ddf.security.oidc.validator.OidcTokenValidator;
+import org.codice.ddf.security.oidc.validator.OidcValidationException;
+import org.codice.ddf.security.token.storage.api.TokenInformation;
+import org.codice.ddf.security.token.storage.api.TokenStorage;
+import org.codice.gsonsupport.GsonTypeAdapters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OAuthSecurityImpl implements OAuthSecurity {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(OAuthSecurityImpl.class);
+
+  static final Gson GSON =
+      new GsonBuilder()
+          .disableHtmlEscaping()
+          .registerTypeAdapterFactory(GsonTypeAdapters.LongDoubleTypeAdapter.FACTORY)
+          .create();
+
+  private static final String CLIENT_CREDENTIALS = "client_credentials";
+  private static final String ACCESS_TOKEN = "access_token";
+  private static final String GRANT_TYPE = "grant_type";
+  private static final String ID_TOKEN = "id_token";
+  private static final String BEARER = "Bearer ";
+  private static final String BASIC = "Basic ";
+  private static final String EXP = "exp";
+
+  private ResourceRetriever resourceRetriever;
+  private TokenStorage tokenStorage;
+
+  public OAuthSecurityImpl(TokenStorage tokenStorage) {
+    this.tokenStorage = tokenStorage;
+    resourceRetriever = new DefaultResourceRetriever();
+  }
+
+  /**
+   * Gets the user's access token from the token storage to set it to the OAUTH header.
+   *
+   * @param client Non-null client to set the access token on.
+   * @param subject subject used to get the user's id (email or username)
+   * @param sourceId the id of the source using OAuth needed to get the correct tokens
```

Comment:
Is sourceId specific to this use case? would this be used for clients not specifically generated for sources?

Reply:
Source ID is the key we when storing tokens. The default file system token storage uses it as a key so we know what client or source the tokens belong to. It doesn't have to be specific to querying a source.